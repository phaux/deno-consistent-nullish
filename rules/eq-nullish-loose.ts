import { getOperatorRange } from "../util/getOperatorRange.ts";

export const eqNullishLooseRule: Deno.lint.Rule = {
  create: (ctx) => ({
    BinaryExpression(exprNode) {
      const op = exprNode.operator;
      if (op !== "===" && op !== "!==") return;

      const nullishNode = getNullishOperand(exprNode);
      if (!nullishNode) return;

      const nullishText = ctx.sourceCode.getText(nullishNode);
      ctx.report({
        range: getOperatorRange(ctx.sourceCode, exprNode),
        message:
          `Strict equality comparison with ${nullishText} is not allowed.`,
        hint: `Differentiating between null and undefined is forbidden. ` +
          `Use \`value ${opMap[op]} ${nullishText}\` ` +
          `to check for any nullish value.`,
        fix(fixer) {
          return fixer.replaceTextRange(
            getOperatorRange(ctx.sourceCode, exprNode),
            opMap[op],
          );
        },
      });
    },
  }),
};

const getNullishOperand = (node: Deno.lint.BinaryExpression) =>
  isNullish(node.right) ? node.right : isNullish(node.left) ? node.left : null;

const isNullish = (node: Deno.lint.Node) =>
  (node.type === "Literal" && node.value == null) ||
  (node.type === "Identifier" && node.name === "undefined");

const opMap = {
  "===": "==",
  "!==": "!=",
} as const;
