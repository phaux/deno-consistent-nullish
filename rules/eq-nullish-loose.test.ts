import { assertSnapshot } from "@std/testing/snapshot";
import { eqNullishLooseRule } from "./eq-nullish-loose.ts";
import { assertEquals } from "@std/assert";

Deno.test("bans strict comparison to null or undefined", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    { name: "test", rules: { rule: eqNullishLooseRule } },
    "test.ts",
    `
    x === null;
    x === undefined;
    x !== null;
    x !== undefined;
    null === x;
    undefined === x;
    null !== x;
    undefined !== x;
    null === undefined;
    undefined !== null;
    null === null;
    undefined !== undefined;
    `
  );
  assertEquals(diagnostics.length, 6); // BUG: Deno parses `!==` as `!=`
  await assertSnapshot(t, diagnostics);
});
