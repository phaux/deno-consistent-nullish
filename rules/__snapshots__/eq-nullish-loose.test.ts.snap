export const snapshot = {};

snapshot[`bans strict comparison to null or undefined 1`] = `
[
  {
    fix: [
      {
        range: [
          7,
          10,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`null\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`null\` is not allowed.",
    range: [
      7,
      10,
    ],
  },
  {
    fix: [
      {
        range: [
          23,
          26,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`undefined\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`undefined\` is not allowed.",
    range: [
      23,
      26,
    ],
  },
  {
    fix: [
      {
        range: [
          84,
          87,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`null\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`null\` is not allowed.",
    range: [
      84,
      87,
    ],
  },
  {
    fix: [
      {
        range: [
          105,
          108,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`undefined\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`undefined\` is not allowed.",
    range: [
      105,
      108,
    ],
  },
  {
    fix: [
      {
        range: [
          158,
          161,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`undefined\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`undefined\` is not allowed.",
    range: [
      158,
      161,
    ],
  },
  {
    fix: [
      {
        range: [
          206,
          209,
        ],
        text: "==",
      },
    ],
    hint: "Differentiating between null and undefined is forbidden. Use \`==\` with \`null\` to check for any nullish value.",
    id: "test/rule",
    message: "Strict equality comparison with \`null\` is not allowed.",
    range: [
      206,
      209,
    ],
  },
]
`;
